[FIX] Precisione decimale non usata durante l'importazione della fattura elettronica

SirAionTech opened the pull request:
Risolve https://github.com/OCA/l10n-italy/issues/4445 per `16.0`.

Sostituisce https://github.com/OCA/l10n-italy/pull/3843 per risolvere https://github.com/OCA/l10n-italy/pull/3843#pullrequestreview-1911224225.
Non ho incluso un test perché il codice che imposta la precisione decimale quando eseguito nei test solleva l'errore:
<details><summary>Stack</summary>
<pre>
  File "/path/to/l10n-italy/l10n_it_fatturapa_in/tests/fatturapa_common.py", line 305, in run_wizard
    return wizard.importFatturaPA()
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/path/to/l10n-italy/l10n_it_fatturapa_in/wizard/wizard_import_fatturapa.py", line 1830, in importFatturaPA
    ) = self._set_decimal_precision("Product Price", "price_decimal_digits")
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/path/to/l10n-italy/l10n_it_fatturapa_in/wizard/wizard_import_fatturapa.py", line 1808, in _set_decimal_precision
    new_cr.commit()
  File "/path/to/odoo/odoo/sql_db.py", line 461, in commit
    self.flush()
  File "/path/to/odoo/odoo/sql_db.py", line 134, in flush
    self.transaction.flush()
  File "/path/to/odoo/odoo/api.py", line 883, in flush
    env_to_flush.flush_all()
  File "/path/to/odoo/odoo/api.py", line 747, in flush_all
    self[model_name].flush_model()
  File "/path/to/odoo/odoo/models.py", line 5630, in flush_model
    self._flush(fnames)
  File "/path/to/odoo/odoo/models.py", line 5731, in _flush
    model.browse(ids)._write(vals)
  File "/path/to/odoo/odoo/models.py", line 3882, in _write
    cr.execute(query, params + [sub_ids])
  File "/path/to/odoo/odoo/sql_db.py", line 321, in execute
    res = self._obj.execute(query, params)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
psycopg2.errors.ForeignKeyViolation: insert or update on table "decimal_precision" violates foreign key constraint "decimal_precision_write_uid_fkey"
DETAIL:  Key (write_uid)=(80) is not present in table "res_users".</pre>
</details> 

----

Ho aggiunto un commit per correggere l'errore nei test https://github.com/OCA/l10n-italy/actions/runs/12313495777/job/34367636978#step:8:1717:
<details><summary>Stack</summary>
<pre>
2024-12-13 10:22:13,191 528 ERROR odoo odoo.addons.l10n_it_fatturapa_pec.tests.test_e_invoice_response: ERROR: TestEInvoiceResponse.test_process_response_RC
Traceback (most recent call last):
  File "/__w/l10n-italy/l10n-italy/l10n_it_fatturapa_pec/tests/test_e_invoice_response.py", line 41, in test_process_response_RC
    ).message_process(False, incoming_mail)
  File "/opt/odoo/addons/mail/models/mail_thread.py", line 1298, in message_process
    routes = self.message_route(message, msg_dict, model, thread_id, custom_values)
  File "/__w/l10n-italy/l10n-italy/l10n_it_fatturapa_pec/models/mail_thread.py", line 77, in message_route
    return self.manage_pec_sdi_notification(message, message_dict)
  File "/__w/l10n-italy/l10n-italy/l10n_it_fatturapa_pec/models/mail_thread.py", line 138, in manage_pec_sdi_notification
    ).create(message_dict)
  File "<decorator-gen-148>", line 2, in create
  File "/opt/odoo/odoo/api.py", line 414, in _model_create_multi
    return create(self, [arg])
  File "/opt/odoo/addons/mail/models/mail_message.py", line 606, in create
    messages = super(Message, self).create(values_list)
  File "<decorator-gen-68>", line 2, in create
  File "/opt/odoo/odoo/api.py", line 415, in _model_create_multi
    return create(self, arg)
  File "/opt/odoo/odoo/addons/base/models/ir_fields.py", line 670, in create
    recs = super().create(vals_list)
  File "<decorator-gen-15>", line 2, in create
  File "/opt/odoo/odoo/api.py", line 415, in _model_create_multi
    return create(self, arg)
  File "/opt/odoo/odoo/models.py", line 3958, in create
    raise ValueError("Invalid field %r on model %r" % (key, self._name))
ValueError: Invalid field 'x_odoo_message_id' on model 'mail.message'</pre>
</details> 

Dovuto al commit https://github.com/odoo/odoo/commit/43041dfd66e984978ad743c4f720e580f82bd1aa che è solo in `16.0`.

eLBati committed:
[FIX] l10n_it_fatturapa_in: Restore original precision after import

Otherwise, if another precision is used during import and an exception is raised, the system precision becomes the precision set during import.

Co-authored-by: Simone Rubino <simone.rubino@aion-tech.it>

SirAionTech committed:
[IMP] l10n_it_fatturapa_in: Computation of prices with many digits

When price precision is increased during import, the price of the created lines should have been computed using the new precision

SirAionTech commented:
@eLBati il primo commit è un cherry-pick del tuo commit https://github.com/OCA/l10n-italy/pull/3843/commits/e7a60c90a70dbee835768805d312b391cb81304f, mi sono aggiunto come co-autore perché l'ho modificato, fammi sapere se per te è ok

MaurizioPellegrinet reviewed (approved):
Test funzionale: OK

MaurizioPellegrinet reviewed (changes requested):
bug
STEPS
1. importo xml con 3 decimali
2. la fattura viene importata correttamente
3. modifica data di registrazione (o conto contropartita)

ATTESO
gli importi dovrebbero rimanere corretti

RISULTATO
alla modifica di data registrazione (o conto contropartita) vengono ricalcolati gli importi
Allego video

https://github.com/user-attachments/assets/f267bda0-78f1-4ac7-89bd-91c760500850

SirAionTech commented:
> bug STEPS
> 
> 1. importo xml con 3 decimali
> 2. la fattura viene importata correttamente
> 3. modifica data di registrazione (o conto contropartita)
> 
> ATTESO gli importi dovrebbero rimanere corretti
> 
> RISULTATO alla modifica di data registrazione (o conto contropartita) vengono ricalcolati gli importi Allego video
> 
>  Screencast.2024-11-21.14.43.38.mp4

Giusto: quando gli importi vengono ricalcolati, si usa la precisione globale non quella _temporanea_ impostata durante l'importazione.

Secondo me è un problema dovuto a queste precisioni _temporanee_ in sé, non alla PR specifica.
Il fatto è che prima non venivano proprio applicate quindi il ricalcolo non cambiava gli importi: aggiustare questa feature in pratica ha fatto scoprire che esiste questo problema.

Comunque ci sta segnalarlo qui, grazie :smile: magari ne parliamo domattina

eLBati commented:
A questo punto vedo 2 strade:
 - quella semplice è togliere del tutto la funzionalità per non confondere l'utente
 - quella completa sarebbe impostare la precisione desiderata a livello di fattura e usarla per tutti i ricalcoli che coinvolgono prezzo, quantità e sconto

Un approccio per il ricalcolo: https://github.com/OCA/l10n-italy/pull/2874

SirAionTech commented:
> A questo punto vedo 2 strade:
> 
> * quella semplice è togliere del tutto la funzionalità per non confondere l'utente
> * quella completa sarebbe impostare la precisione desiderata a livello di fattura e usarla per tutti i ricalcoli che coinvolgono prezzo, quantità e sconto
> 
> Un approccio per il ricalcolo: #2874

Ok grazie, vedo come fare

SirAionTech commented:
I test di **l10n_it_financial_statements_report** falliscono per la modifica https://github.com/OCA/account-financial-reporting/pull/1243 mergiata poco fa nella sua dipendenza **account_financial_report**.

SirAionTech commented:
/ocabot rebase

MaurizioPellegrinet reviewed (approved):
Test funzionale: OK

TheMule71 commented:
/ocabot merge minor

TheMule71 reviewed (approved):
LGTM

SirAionTech commented:
> @TheMule71 your merge command was aborted due to failed check(s), which you can inspect on [this commit of 16.0-ocabot-merge-pr-4446-by-TheMule71-bump-minor](https://github.com/OCA/l10n-italy/commits/f321c6aeed6743f9a9c0ee4e34875b2271717ff4).
> 
> After fixing the problem, you can re-issue a merge command. Please refrain from merging manually as it will most probably make the target branch red.

3 giorni fa era tutto :green_circle:
:thinking: vediamo come va con un rebase
/ocabot rebase

SirAionTech committed:
[FIX] l10n_it_fatturapa_pec: Remove extra fields from message

`message_dict` contains values used to track the message, but we use it to create a `mail.message` so the extra values have to be removed because they are not fields of `mail.message`.

TheMule71 commented:
/ocabot merge minor
